### PR TITLE
Refactor Agent to use DialogueKit

### PR DESCRIPTION
**Pull request by NoB0**
## What's changed?

- The class `Agent` was renamed to `MovieBotAgent` and inherits from `dialoguekit.participant.Agent`.
- The class `MovieBotDialogueConnector` inherits from `dialoguekit.connector.DialogueConnector` and handles the connection between agents and users.
- The class `DialogueContext` was removed. The previous utterances can be accessed directly in the dialogue manager. The list of movies recommended is now part of `DialogueState`.
- The class `Controller` inherits from `dialoguekit.platforms.Platform`.

Follow up includes the creation of tests for the class `MovieBotAgent`.

Depends on #167

### CODE REVIEW

File: moviebot/agent/agent.py [hunk 1]
```diff
@@ -142,112 +135,117 @@ def _get_recommender(self, recommender_type: str) -> RecommenderModel:
...
-        Returns:
-            Agent response, dialogue options for quick reply, and data to be
-              recorded if self.isBot is true.
+        utterance = self._generate_utterance(
+            agent_response, AgentIntents.WELCOME.value, options
+        )
+
+        self._dialogue_connector.register_agent_utterance(utterance)
+
+    def goodbye(self) -> None:
```
Comment (IKostric) [inline moviebot/agent/agent.py:199]: I don't think this method is in use. Dialogue policy generates `bye` intent, and dialogue connector closes the conversation.

**Review by IKostric — APPROVED** [review]
LGTM,
Couple of nits.

File: moviebot/connector/dialogue_connector.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        self._agent.receive_utterance(user_utterance, user_options)
+
+    def close(self) -> None:
+        """Closes the conversation."""
+        if self._save_dialogue_history:
+            self._stringify_dialogue_acts()
+            self._dump_dialogue_history()
+
+    def _stringify_dialogue_acts(self) -> None:
+        """Stringifies the dialogue acts before dump."""
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/connector/dialogue_connector.py:67]: Good point, done.

File: moviebot/controller/controller.py [hunk 3]
```diff
@@ -5,44 +5,63 @@
...
 
 
-class Controller(ABC):
-    def __init__(self, configuration: Dict[str, Any]):
-        """This is the main class that controls the other components of the
-        IAI MovieBot. The controller executes the conversational agent.
+class Controller(DialogueKitPlatform, ABC):
+    def __init__(
+        self,
+        agent_class: Type["MovieBotAgent"],
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/controller/controller.py:30]: Done.

File: moviebot/controller/controller.py [hunk 4]
```diff
@@ -5,44 +5,63 @@
...
+        super().__init__(agent_class)
+
+        self._config = config
+        self._active_users: DefaultDict[str, User] = defaultdict(User)
 
-    @abstractmethod
-    def execute_agent(self):
-        """Runs the conversational agent and executes the dialogue by calling
-        the basic components of IAI MovieBot.
+    def get_new_agent(self) -> "MovieBotAgent":
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/controller/controller.py:44]: Done.

File: moviebot/controller/controller.py [hunk 5]
```diff
@@ -5,44 +5,63 @@
...
 
-        Raises:
-            NotImplementedError: If the method is not implemented.
+        Returns:
+            Agent.
         """
-        raise NotImplementedError
+        return self._agent_class(**self._config)
+
+    def connect(self, user_id: str) -> None:
```
[the remark this answers is not in the document]
Comment (NoB0) [reply moviebot/controller/controller.py:52]: Unlike the parent method, this one instantiates a `MovieBotDialogueConnector`.

File: moviebot/agent/agent.py (hunk 1 above)
Comment (NoB0) [reply moviebot/agent/agent.py:199]: Good catch. However, the method needs to be implemented as it is an abstract method in the parent class.

**Pull request merged**